java/src/jmri/jmrit/operations/rollingstock/RollingStock.java

```diff
@@ -813,7 +815,7 @@ public void propertyChange(java.beans.PropertyChangeEvent e) {
                         if (e.getNewValue() != null) {
                             Location newLocation = locationManager.getLocationByReporter((jmri.Reporter) e.getNewValue());
                             if (newLocation != getLocation())
-                                setLocation(newLocation, null);
+                                setWhereLastSeen(newLocation);
```

Comment:
Paul,

 

Thanks for getting back to me.  I don’t have any experience with readers or tags, so you’re doing an excellent job of educating me.  The user believes that tags will help him figure out where his misplaced cars are on his layout.  He also want to know how his cars are blocked in staging.  He’s a long time operations user and I’m trying to help him. More below.

 

Dan

 

From: Paul Bender [mailto:notifications@github.com] 
Sent: Wednesday, August 10, 2016 7:31 PM
To: JMRI/JMRI <JMRI@noreply.github.com>
Cc: Daniel Boudreau <daboudreau@hotmail.com>; State change <state_change@noreply.github.com>
Subject: Re: [JMRI/JMRI] Add to rolling stock RFID "where last seen" and "when last seen" (#1767)

 

In java/src/jmri/jmrit/operations/rollingstock/RollingStock.java <https://github.com/JMRI/JMRI/pull/1767#discussion_r74350177> :

> @@ -813,7 +815,7 @@ public void propertyChange(java.beans.PropertyChangeEvent e) {
>                          if (e.getNewValue() != null) {
>                              Location newLocation = locationManager.getLocationByReporter((jmri.Reporter) e.getNewValue());
>                              if (newLocation != getLocation())
> -                                setLocation(newLocation, null);
> +                                setWhereLastSeen(newLocation);

Sent from my iPad

On Aug 10, 2016, at 5:20 PM, Daniel Boudreau ***@***.*** <mailto:***@***.***> > wrote: Hi Paul, Sorry if I messed up your scripts. The issue is that I have a user that wants to comparet what the operations program believes the car is, versus what is actually provided by readers. Previous to my latest change, when a car went through a reader, the code would update the car’s location, and null out the car’s track. Then if the operations program detected train movement, would override the reader location, and update the car’s track. 

Correct, and that's the correct behavior when a reader is located on the mainline or an arrival/departure track in a yard.

The user asked that the operations car’s xml file provide him with both the location that operation believes the car is and what the reporters registered. He wants to write his own program to find those discrepancies using the operations xml car file. I was planning on adding more code that would highlight all cars in the table where a car’s operation location doesn’t match the tag’s last known location. I’m also planning on logging all the “where last seen” location changes so a user can review how the car got to a particular location. 

If you only have readers at key points ( entering and exiting yards and at junction points )

 

I agree that there are most likely a limited number of readers on the layout.  We know the cars route, we also know where the readers are, so we should be able to figure out if the car isn’t at the right location.  There could be a route where a car never passes a reader, or a reader that is several locations behind where the car eventually ends.  We can also flag a car passes a reader that it shouldn’t.

 

> 1) Why are you duplicating the When Last Seen and Where Last Seen properties of the ID Tag? > > Those should be persistent values associated with the ID tag, so we can look those up, and > > shouldn't need to be stored with the rolling stock. > My bad. Didn’t realize that the tag was keeping those two things. I see that now. The user also wants to have a reader for each track in staging so he can have a wheel report by track. So I’ll either need to save this info in operations, or add it to the ID tag. 

That's actually overkill. You can build a wheel report by track with a single reader at the yard thought using switch position information to determine

 

That’s great way of doing it, is there something already coded that maps switch positions to a track?  I want to pass this along to him.

 

> 2) The code that sets Where Last Seen and When Last Seen instead of the Location and the > > Date also need to store the Location and the Date. We may not have ID tag readers at ever > > location, but when a car passes a reader, that reader is associated with a location, and we > > know the car MUST be at that location. > I agree. But the two fields were being updated by two very different mechanisms. 

No. The location and date were both updated when the reader sees a tag and notified the reader that either the last. Yes, the location also changes when the train location is updated by the user, but that is after the point at which the information from the readers is useful. The readers tell us that a piece of rolling stock passed a location at a given time. By the time the user moves the train to a new location, the expectation is that the car order will have changed, so we can't rely on the old tag generated location and date to be correct.

 

This user is also using CATs to update train locations.  Pretty sure that either the reader or the train update is going to override the location or last seen date.

 

Also let’s also say the car is on the wrong train., the reader detects the car, and then later the train that should have the car moves, the reader info is lost, and we don’t know where the car is, unless we have the wheel report.  Do you always create a wheel report immediately after a train passes a reader?  I see a race condition between two update mechanisms, does that make sense to you?

 

 

> The 2nd part of this is VERY important to me, because it changes the previous behaviour. > > This code change breaks scripts that I've got locally to print wheel reports. > Are you using operations and readers at the same time? I don’t understand how this would work. I would think if the ID tag has the car’s last known location and date, that your scripts could pull this from the car’s ID tag and not from the operation’s database. 

 

Yes, I am using operations with readers. Remember, I'm the one that put The code in place to update the locations to begin with. I am actually replicating the information that RFID readers provide the prototype, which is that a piece of rolling stock passed a reader at a key point location. When the train moves ( via either the move or the terminate button ) that represents another operation on the prototype ( which is a clerk making a notation in the computer that a train has moved ). More importantly though is that you changed the behavior of something that has been in production and at least a few people have come to use. We have generally been trying to avoid changing the expected behavior of the API at a moments notice, and this change definitely does that. Incidentally, one of the projects I have planned is to also be able to associate an ID tag with a train. The ID tags will be functionally replace OSing the train location, and will be used to move the train along the route. For my own purposes, the ID tag will be credit card sized RFID swiped on a fascia mounted reader, but there will be options to use the tag from a piece of rolling stock in the train ( cabooses, locomotives,etc ) as well. The other thing project that goes along with the fascia mounted RFID readers to update train positions is a fascia mounted receipt printer that will print out the wheel reports and/or switchlists at key locations. I'm in the process of installing one of these at a new staging yard I started building recently. Paul

 

I agree that I should break existing code, but I’m thinking that it’s broke the way it was written and I’m trying to fix it.  Given that entering a car’s tag Id  from operations has been broken for a while, I don’t think we have a lot of users using tags and operations.

 

Not sure when and how your script creates the wheel report, but if it doesn’t react immediately to a car location change, there’s a chance that the train move update will overwrite the location and date.   I believe there’s a race condition that needs to be addressed.

 

Dan

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub <https://github.com/JMRI/JMRI/pull/1767/files/b8a4457a349fa99fa532337fd1e13bf1b2ad22c2#r74350177> , or mute the thread <https://github.com/notifications/unsubscribe-auth/ANw2p9GoF8TCP89pXN8UU93dk5UXdVT4ks5qel8ggaJpZM4JhdU1> .  <https://github.com/notifications/beacon/ANw2pzXK20e9Son58IH_qJti3B8t5kspks5qel8ggaJpZM4JhdU1.gif>